File: app/src/main/java/com/san_pedrito/myapplication/db_kt/LaptopDatabaseHelper.java
package com.san_pedrito.myapplication.db_kt;

import android.content.ContentValues;
import android.content.Context;
import android.database.Cursor;
import android.database.sqlite.SQLiteDatabase;
import android.database.sqlite.SQLiteOpenHelper;
import java.util.ArrayList;
import java.util.List;

public class LaptopDatabaseHelper extends SQLiteOpenHelper {
    private static final String DATABASE_NAME = "laptops.db";
    private static final int DATABASE_VERSION = 1;

    public LaptopDatabaseHelper(Context context) {
        super(context, DATABASE_NAME, null, DATABASE_VERSION);
    }

    @Override
    public void onCreate(SQLiteDatabase db) {
        String CREATE_LAPTOPS_TABLE = "CREATE TABLE laptops (" +
                "id INTEGER PRIMARY KEY AUTOINCREMENT," +
                "marca TEXT," +
                "modelo TEXT," +
                "numero_serie TEXT," +
                "estado TEXT," +
                "observaciones TEXT," +
                "ruta_imagen TEXT," +
                "fecha_hora DATETIME DEFAULT CURRENT_TIMESTAMP)";
        db.execSQL(CREATE_LAPTOPS_TABLE);
    }

    @Override
    public void onUpgrade(SQLiteDatabase db, int oldVersion, int newVersion) {
        db.execSQL("DROP TABLE IF EXISTS laptops");
        onCreate(db);
    }

    public long agregarLaptop(String marca, String modelo, String numeroSerie, 
                            String estado, String observaciones, String rutaImagen) {
        SQLiteDatabase db = this.getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put("marca", marca);
        values.put("modelo", modelo);
        values.put("numero_serie", numeroSerie);
        values.put("estado", estado);
        values.put("observaciones", observaciones);
        values.put("ruta_imagen", rutaImagen);
        return db.insert("laptops", null, values);
    }

    public List<Laptop> obtenerTodasLaptops() {
        List<Laptop> laptops = new ArrayList<>();
        SQLiteDatabase db = this.getReadableDatabase();
        Cursor cursor = db.query("laptops", null, null, null, null, null, "fecha_hora DESC");

        if (cursor.moveToFirst()) {
            do {
                Laptop laptop = new Laptop(
                    cursor.getLong(cursor.getColumnIndexOrThrow("id")),
                    cursor.getString(cursor.getColumnIndexOrThrow("marca")),
                    cursor.getString(cursor.getColumnIndexOrThrow("modelo")),
                    cursor.getString(cursor.getColumnIndexOrThrow("numero_serie")),
                    cursor.getString(cursor.getColumnIndexOrThrow("estado")),
                    cursor.getString(cursor.getColumnIndexOrThrow("observaciones")),
                    cursor.getString(cursor.getColumnIndexOrThrow("ruta_imagen")),
                    cursor.getString(cursor.getColumnIndexOrThrow("fecha_hora"))
                );
                laptops.add(laptop);
            } while (cursor.moveToNext());
        }
        cursor.close();
        return laptops;
    }

    public Laptop obtenerLaptopPorNumeroSerie(String numeroSerie) {
        SQLiteDatabase db = this.getReadableDatabase();
        Cursor cursor = db.query("laptops", null, "numero_serie = ?", 
            new String[]{numeroSerie}, null, null, null);

        Laptop laptop = null;
        if (cursor.moveToFirst()) {
            laptop = new Laptop(
                cursor.getLong(cursor.getColumnIndexOrThrow("id")),
                cursor.getString(cursor.getColumnIndexOrThrow("marca")),
                cursor.getString(cursor.getColumnIndexOrThrow("modelo")),
                cursor.getString(cursor.getColumnIndexOrThrow("numero_serie")),
                cursor.getString(cursor.getColumnIndexOrThrow("estado")),
                cursor.getString(cursor.getColumnIndexOrThrow("observaciones")),
                cursor.getString(cursor.getColumnIndexOrThrow("ruta_imagen")),
                cursor.getString(cursor.getColumnIndexOrThrow("fecha_hora"))
            );
        }
        cursor.close();
        return laptop;
    }

    public int actualizarLaptop(long id, String marca, String modelo, String numeroSerie,
                               String estado, String observaciones, String rutaImagen) {
        SQLiteDatabase db = this.getWritableDatabase();
        ContentValues values = new ContentValues();
        values.put("marca", marca);
        values.put("modelo", modelo);
        values.put("numero_serie", numeroSerie);
        values.put("estado", estado);
        values.put("observaciones", observaciones);
        values.put("ruta_imagen", rutaImagen);
        return db.update("laptops", values, "id = ?", new String[]{String.valueOf(id)});
    }

    public void eliminarLaptop(long id) {
        SQLiteDatabase db = this.getWritableDatabase();
        db.delete("laptops", "id = ?", new String[]{String.valueOf(id)});
    }

    public boolean existeNumeroSerie(String numeroSerie) {
        SQLiteDatabase db = this.getReadableDatabase();
        Cursor cursor = db.query(
            "laptops",
            new String[]{"numero_serie"},
            "numero_serie = ?",
            new String[]{numeroSerie},
            null,
            null,
            null
        );
        boolean existe = cursor.getCount() > 0;
        cursor.close();
        return existe;
    }
} 

File: app/src/main/java/com/san_pedrito/myapplication/db_kt/LaptopDatabaseHelper.kt
package com.san_pedrito.myapplication.db_kt

import android.content.ContentValues
import android.content.Context
import android.database.sqlite.SQLiteDatabase
import android.database.sqlite.SQLiteOpenHelper

/**
 * Clase auxiliar para gestionar la base de datos de laptops.
 * Proporciona métodos para crear, leer, actualizar y eliminar registros de laptops.
 */
class LaptopDatabaseHelper(context: Context) : SQLiteOpenHelper(context, DATABASE_NAME, null, DATABASE_VERSION) {

    companion object {
        private const val DATABASE_VERSION = 2
        private const val DATABASE_NAME = "LaptopsDB"
        private const val TABLA_LAPTOPS = "laptops"
        private const val CLAVE_ID = "id"
        private const val CLAVE_MARCA = "marca"
        private const val CLAVE_MODELO = "modelo"
        private const val CLAVE_NUMERO_SERIE = "serial_number"
        private const val CLAVE_ESTADO = "estado"
        private const val CLAVE_OBSERVACIONES = "observaciones"
        private const val CLAVE_RUTA_IMAGEN = "image_path"
        private const val CLAVE_FECHA_HORA = "timestamp"
    }

    /**
     * Crea la tabla de laptops en la base de datos.
     */
    override fun onCreate(db: SQLiteDatabase) {
        val CREAR_TABLA_LAPTOPS = ("CREATE TABLE " + TABLA_LAPTOPS + "("
                + CLAVE_ID + " INTEGER PRIMARY KEY,"
                + CLAVE_MARCA + " TEXT,"
                + CLAVE_MODELO + " TEXT,"
                + CLAVE_NUMERO_SERIE + " TEXT,"
                + CLAVE_ESTADO + " TEXT,"
                + CLAVE_OBSERVACIONES + " TEXT,"
                + CLAVE_RUTA_IMAGEN + " TEXT,"
                + CLAVE_FECHA_HORA + " TEXT" + ")")
        db.execSQL(CREAR_TABLA_LAPTOPS)
    }

    /**
     * Actualiza la estructura de la base de datos cuando cambia la versión.
     */
    override fun onUpgrade(db: SQLiteDatabase, oldVersion: Int, newVersion: Int) {
        if (oldVersion < 2) {
            // Añadir columna de fecha y hora a la tabla existente
            db.execSQL("ALTER TABLE $TABLA_LAPTOPS ADD COLUMN $CLAVE_FECHA_HORA TEXT")
        } else {
            // Recrear la tabla como alternativa
            db.execSQL("DROP TABLE IF EXISTS $TABLA_LAPTOPS")
            onCreate(db)
        }
    }

    /**
     * Añade una nueva laptop a la base de datos.
     * @return ID del registro insertado
     */
    fun agregarLaptop(marca: String, modelo: String, numeroSerie: String, estado: String, observaciones: String, rutaImagen: String?): Long {
        val db = this.writableDatabase
        val values = ContentValues()
        values.put(CLAVE_MARCA, marca)
        values.put(CLAVE_MODELO, modelo)
        values.put(CLAVE_NUMERO_SERIE, numeroSerie)
        values.put(CLAVE_ESTADO, estado)
        values.put(CLAVE_OBSERVACIONES, observaciones)
        values.put(CLAVE_RUTA_IMAGEN, rutaImagen)
        
        // Añadir fecha y hora actual
        val fechaHora = java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss", java.util.Locale.getDefault()).format(java.util.Date())
        values.put(CLAVE_FECHA_HORA, fechaHora)
        
        return db.insert(TABLA_LAPTOPS, null, values)
    }

    /**
     * Obtiene una laptop por su ID.
     * @return Objeto Laptop o null si no se encuentra
     */
    fun obtenerLaptop(id: Long): Laptop? {
        val db = this.readableDatabase
        val cursor = db.query(
            TABLA_LAPTOPS, arrayOf(CLAVE_ID, CLAVE_MARCA, CLAVE_MODELO, CLAVE_NUMERO_SERIE, CLAVE_ESTADO, CLAVE_OBSERVACIONES, CLAVE_RUTA_IMAGEN, CLAVE_FECHA_HORA),
            "$CLAVE_ID=?", arrayOf(id.toString()), null, null, null, null)
        return if (cursor.moveToFirst()) {
            val laptopId = cursor.getLong(cursor.getColumnIndexOrThrow(CLAVE_ID))
            val marca = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_MARCA))
            val modelo = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_MODELO))
            val numeroSerie = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_NUMERO_SERIE))
            val estado = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_ESTADO))
            val observaciones = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_OBSERVACIONES))
            val rutaImagen = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_RUTA_IMAGEN))
            val fechaHora = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_FECHA_HORA))
            cursor.close()
            Laptop(laptopId, marca, modelo, numeroSerie, estado, observaciones, rutaImagen, fechaHora)
        } else {
            cursor.close()
            null
        }
    }

    /**
     * Obtiene todas las laptops de la base de datos.
     * @return Lista de objetos Laptop
     */
    fun obtenerTodasLaptops(): List<Laptop> {
        val listaLaptops = mutableListOf<Laptop>()
        val consultaSeleccion = "SELECT * FROM $TABLA_LAPTOPS"
        val db = this.readableDatabase
        val cursor = db.rawQuery(consultaSeleccion, null)
        if (cursor.moveToFirst()) {
            do {
                val id = cursor.getLong(cursor.getColumnIndexOrThrow(CLAVE_ID))
                val marca = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_MARCA))
                val modelo = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_MODELO))
                val numeroSerie = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_NUMERO_SERIE))
                val estado = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_ESTADO))
                val observaciones = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_OBSERVACIONES))
                val rutaImagen = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_RUTA_IMAGEN))
                val fechaHora = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_FECHA_HORA))
                listaLaptops.add(Laptop(id, marca, modelo, numeroSerie, estado, observaciones, rutaImagen, fechaHora))
            } while (cursor.moveToNext())
        }
        cursor.close()
        return listaLaptops
    }

    /**
     * Actualiza la información de una laptop existente.
     * @return Número de filas afectadas
     */
    fun actualizarLaptop(id: Long, marca: String, modelo: String, numeroSerie: String, estado: String, observaciones: String, rutaImagen: String?): Int {
        val db = this.writableDatabase
        val values = ContentValues()
        values.put(CLAVE_MARCA, marca)
        values.put(CLAVE_MODELO, modelo)
        values.put(CLAVE_NUMERO_SERIE, numeroSerie)
        values.put(CLAVE_ESTADO, estado)
        values.put(CLAVE_OBSERVACIONES, observaciones)
        values.put(CLAVE_RUTA_IMAGEN, rutaImagen)
        
        // Añadir fecha y hora actual al actualizar
        val fechaHora = java.text.SimpleDateFormat("yyyy-MM-dd HH:mm:ss", java.util.Locale.getDefault()).format(java.util.Date())
        values.put(CLAVE_FECHA_HORA, fechaHora)
        
        return db.update(TABLA_LAPTOPS, values, "$CLAVE_ID=?", arrayOf(id.toString()))
    }

    /**
     * Elimina una laptop de la base de datos.
     * @return Número de filas afectadas
     */
    fun eliminarLaptop(id: Long): Int {
        val db = this.writableDatabase
        return db.delete(TABLA_LAPTOPS, "$CLAVE_ID=?", arrayOf(id.toString()))
    }

    /**
     * Elimina todas las laptops de la base de datos.
     * @return Número de filas afectadas
     */
    fun eliminarTodasLaptops(): Int {
        val db = this.writableDatabase
        return db.delete(TABLA_LAPTOPS, null, null)
    }

    /**
     * Busca laptops que coincidan con la consulta en marca, modelo o número de serie.
     * @return Lista de objetos Laptop que coinciden con la búsqueda
     */
    fun buscarLaptops(consulta: String): List<Laptop> {
        val listaLaptops = mutableListOf<Laptop>()
        val consultaSeleccion = "SELECT * FROM $TABLA_LAPTOPS WHERE $CLAVE_MARCA LIKE '%$consulta%' OR $CLAVE_MODELO LIKE '%$consulta%' OR $CLAVE_NUMERO_SERIE LIKE '%$consulta%'"
        val db = this.readableDatabase
        val cursor = db.rawQuery(consultaSeleccion, null)
        if (cursor.moveToFirst()) {
            do {
                val id = cursor.getLong(cursor.getColumnIndexOrThrow(CLAVE_ID))
                val marca = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_MARCA))
                val modelo = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_MODELO))
                val numeroSerie = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_NUMERO_SERIE))
                val estado = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_ESTADO))
                val observaciones = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_OBSERVACIONES))
                val rutaImagen = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_RUTA_IMAGEN))
                val fechaHora = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_FECHA_HORA))
                listaLaptops.add(Laptop(id, marca, modelo, numeroSerie, estado, observaciones, rutaImagen, fechaHora))
            } while (cursor.moveToNext())
        }
        cursor.close()
        return listaLaptops
    }

    /**
     * Obtiene una laptop por su número de serie.
     * @return Objeto Laptop o null si no se encuentra
     */
    fun obtenerLaptopPorNumeroSerie(numeroSerie: String): Laptop? {
        val db = this.readableDatabase
        val cursor = db.query(
            TABLA_LAPTOPS, arrayOf(CLAVE_ID, CLAVE_MARCA, CLAVE_MODELO, CLAVE_NUMERO_SERIE, CLAVE_ESTADO, CLAVE_OBSERVACIONES, CLAVE_RUTA_IMAGEN, CLAVE_FECHA_HORA),
            "$CLAVE_NUMERO_SERIE=?", arrayOf(numeroSerie), null, null, null, null)
        return if (cursor.moveToFirst()) {
            val laptopId = cursor.getLong(cursor.getColumnIndexOrThrow(CLAVE_ID))
            val marca = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_MARCA))
            val modelo = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_MODELO))
            val serie = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_NUMERO_SERIE))
            val estado = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_ESTADO))
            val observaciones = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_OBSERVACIONES))
            val rutaImagen = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_RUTA_IMAGEN))
            val fechaHora = cursor.getString(cursor.getColumnIndexOrThrow(CLAVE_FECHA_HORA))
            cursor.close()
            Laptop(laptopId, marca, modelo, serie, estado, observaciones, rutaImagen, fechaHora)
        } else {
            cursor.close()
            null
        }
    }
}